cookbooks/aws-parallelcluster-byos/resources/create_user.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+# Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+# with the License. A copy of the License is located at http://aws.amazon.com/apache2.0/
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+resource_name :create_user
+provides :create_user
+
+# Resource to create system user
+
+property :name, String, name_property: true
+property :system_users, Array, required: false
+
+default_action :run
+
+def check_gid(gid)
+  check_group_stdout = Mixlib::ShellOut.new("getent group #{gid}").run_command.stdout.strip
+  return unless check_group_stdout.empty?
+
+  raise("gid #{gid} is used by #{check_group_stdout}, it should be reserved for ParallelCluster system group. " \
+        "Reserved gid range is #{node['cluster']['byos']['system_group_id_start']}-" \
+        "#{node['cluster']['byos']['system_group_id_start'] + 9}.")
+end
+
+def check_uid(uid)
+  check_user_stdout = Mixlib::ShellOut.new("getent passwd #{uid}").run_command.stdout.strip
+  return unless check_user_stdout.empty?
+
+  raise("uid #{uid} is used by #{check_user_stdout}, it should be reserved for ParallelCluster system user. " \
+        "Reserved uid range is #{node['cluster']['byos']['system_user_id_start']}-" \
+        "#{node['cluster']['byos']['system_user_id_start'] + 9}.")
+end
+
+action :run do
+  Chef::Log.debug("Called create_user with system users (#{new_resource.system_users})")
+
+  if new_resource.system_users && !new_resource.system_users.empty?
+
+    new_resource.system_users.each_with_index do |user, index|
+      name = user[:Name]
+      enable_imds = user[:EnableImds]
+      system_user_id = node['cluster']['byos']['system_user_id_start'] + index
+      system_group_id = node['cluster']['byos']['system_group_id_start'] + index
+
+      check_gid(system_group_id)
+
+      Chef::Log.info("create gid #{system_group_id} user with group name #{name}.")
+      group name do
+        comment "byos system group #{name}"
+        gid system_group_id
+        system true
+      end
+
+      check_uid(system_user_id)
```

Comment:
minor - this can be moved next to the check_gid, so that first checks are done then user is created

Reply:
done